File: aubo_ros2_ws/src/coordinate_transforms/src/core/point_transform.cpp
#include "coordinate_transforms/core/point_transform.hpp"
#include <cmath>

namespace coordinate_transforms {
namespace core {

// 手系变换矩阵：右手系↔左手系。右手系 X×Y=Z，左手系 X×Y=-Z。
// 公式: 仅旋转时用反射矩阵，如 M = diag(1, 1, -1) 使 Z 取反，即 [x,y,z]' -> [x,y,-z]。
Matrix3 handedness_transform(Handedness from, Handedness to) {
  if (from == to) {
    return {1,0,0, 0,1,0, 0,0,1};
  }
  return {1,0,0, 0,1,0, 0,0,-1};
}

Matrix3 right_to_left() {
  return handedness_transform(Handedness::Right, Handedness::Left);
}

Matrix3 left_to_right() {
  return handedness_transform(Handedness::Left, Handedness::Right);
}

// 旋转: p' = R*p（行优先 R[0..8] 对应 3×3）。
static Point3 apply_R(const Point3& p, const Matrix3& R) {
  return {
    R[0]*p[0] + R[1]*p[1] + R[2]*p[2],
    R[3]*p[0] + R[4]*p[1] + R[5]*p[2],
    R[6]*p[0] + R[7]*p[1] + R[8]*p[2]
  };
}

// 刚体变换: p' = T*[p;1] 取前三维，即 p' = R*p + t。
static Point3 apply_T(const Point3& p, const Matrix4& T) {
  return {
    T[0]*p[0] + T[1]*p[1] + T[2]*p[2] + T[3],
    T[4]*p[0] + T[5]*p[1] + T[6]*p[2] + T[7],
    T[8]*p[0] + T[9]*p[1] + T[10]*p[2] + T[11]
  };
}

// 单点变换: p' = T*p（齐次），等价 p' = R*p + t。
Point3 transform_point(const Point3& p, const Matrix4& T) {
  return apply_T(p, T);
}

// 单点变换: p' = R*p + t。
Point3 transform_point(const Point3& p, const Matrix3& R, const Point3& t) {
  Point3 rp = apply_R(p, R);
  return {rp[0]+t[0], rp[1]+t[1], rp[2]+t[2]};
}

// 方向向量变换: v' = R*v（仅旋转，无平移）；可选单位化 v' := v'/|v'|。
Point3 transform_direction(const Point3& v, const Matrix3& R, bool normalize) {
  Point3 out = apply_R(v, R);
  if (normalize) {
    double n = std::sqrt(out[0]*out[0] + out[1]*out[1] + out[2]*out[2]);
    if (n > 1e-10) { out[0]/=n; out[1]/=n; out[2]/=n; }
  }
  return out;
}

Point3 transform_direction(const Point3& v, const Matrix4& T, bool normalize) {
  Matrix3 R = {T[0],T[1],T[2], T[4],T[5],T[6], T[8],T[9],T[10]};
  return transform_direction(v, R, normalize);
}

// 多点变换: 对每点 p_i 做 p'_i = T*p_i（齐次），即 P' = (T * P^T)^T 取前 3 列。
std::vector<Point3> transform_points(const std::vector<Point3>& P, const Matrix4& T) {
  std::vector<Point3> out(P.size());
  for (size_t i = 0; i < P.size(); ++i) out[i] = apply_T(P[i], T);
  return out;
}

// 位姿/架变换: 原点 origin' = T_B_A*origin；姿态 new_R = R_B_A * R（矩阵乘）。
// 即完整架 T_old = [R|origin; 0 0 0 1]，变换后 T_new = T_B_A * T_old。
void transform_pose(const Point3& origin, const Matrix3& R, const Matrix4& T_B_A,
                    Point3& new_origin, Matrix3& new_R) {
  new_origin = apply_T(origin, T_B_A);
  Matrix3 R_B_A = {T_B_A[0],T_B_A[1],T_B_A[2], T_B_A[4],T_B_A[5],T_B_A[6], T_B_A[8],T_B_A[9],T_B_A[10]};
  new_R[0] = R_B_A[0]*R[0] + R_B_A[1]*R[3] + R_B_A[2]*R[6];
  new_R[1] = R_B_A[0]*R[1] + R_B_A[1]*R[4] + R_B_A[2]*R[7];
  new_R[2] = R_B_A[0]*R[2] + R_B_A[1]*R[5] + R_B_A[2]*R[8];
  new_R[3] = R_B_A[3]*R[0] + R_B_A[4]*R[3] + R_B_A[5]*R[6];
  new_R[4] = R_B_A[3]*R[1] + R_B_A[4]*R[4] + R_B_A[5]*R[7];
  new_R[5] = R_B_A[3]*R[2] + R_B_A[4]*R[5] + R_B_A[5]*R[8];
  new_R[6] = R_B_A[6]*R[0] + R_B_A[7]*R[3] + R_B_A[8]*R[6];
  new_R[7] = R_B_A[6]*R[1] + R_B_A[7]*R[4] + R_B_A[8]*R[7];
  new_R[8] = R_B_A[6]*R[2] + R_B_A[7]*R[5] + R_B_A[8]*R[8];
}

// 完整架变换: T_new = T_B_A * T_old（4×4 矩阵乘）。同一刚体变换作用到架的原点与姿态。
Matrix4 transform_frame(const Matrix4& T_old, const Matrix4& T_B_A) {
  Matrix4 out{};
  for (int i = 0; i < 4; ++i)
    for (int j = 0; j < 4; ++j) {
      out[i*4+j] = 0;
      for (int k = 0; k < 4; ++k)
        out[i*4+j] += T_B_A[i*4+k] * T_old[k*4+j];
    }
  return out;
}

// 多方向向量: 每轴 v'_i = R*v_i，可选单位化。
std::vector<Point3> transform_directions(const std::vector<Point3>& V, const Matrix3& R, bool normalize) {
  std::vector<Point3> out(V.size());
  for (size_t i = 0; i < V.size(); ++i) out[i] = transform_direction(V[i], R, normalize);
  return out;
}

// 直线（一点+方向）变换: point' = T*point；direction' = R*direction 并单位化（方向仅旋转）。
void transform_line(const Point3& point, const Point3& direction, const Matrix4& T,
                    Point3& point_out, Point3& direction_out) {
  point_out = apply_T(point, T);
  Matrix3 R = {T[0],T[1],T[2], T[4],T[5],T[6], T[8],T[9],T[10]};
  direction_out = transform_direction(direction, R, true);
}

}  // namespace core
}  // namespace coordinate_transforms

File: aubo_ros2_ws/src/coordinate_transforms_py/coordinate_transforms_py/core/point_transform.py
# Core: point, direction, batch, frame, line, handedness (no ROS)，使用 NumPy 做矩阵/向量运算
import numpy as np
from typing import Tuple
from enum import Enum


class Handedness(Enum):
    Right = 0
    Left = 1


def _ensure_3(p) -> np.ndarray:
    return np.asarray(p, dtype=np.float64).reshape(3)


def _ensure_4x4(T) -> np.ndarray:
    return np.asarray(T, dtype=np.float64).reshape(4, 4)


def _ensure_3x3(M) -> np.ndarray:
    return np.asarray(M, dtype=np.float64).reshape(3, 3)


# 手系变换：右手系↔左手系。右手系 X×Y=Z，左手系 X×Y=-Z。公式: 反射矩阵 M = diag(1,1,-1)，[x,y,z]' → [x,y,-z]
def handedness_transform(from_h: Handedness, to_h: Handedness) -> np.ndarray:
    if from_h == to_h:
        return np.eye(3)
    return np.diag([1.0, 1.0, -1.0])

def right_to_left() -> np.ndarray:
    return handedness_transform(Handedness.Right, Handedness.Left)

def left_to_right() -> np.ndarray:
    return handedness_transform(Handedness.Left, Handedness.Right)

# 刚体变换单点: p' = T*[p;1] 取前三维。公式: p' = R*p + t，T = [R|t; 0 0 0 1] 行优先
def transform_point(p: np.ndarray, T: np.ndarray) -> np.ndarray:
    p = _ensure_3(p)
    T = _ensure_4x4(T)
    ph = np.append(p, 1.0)
    out = T @ ph
    return out[:3]


# 方向向量变换: v' = R*v（仅旋转）；可选单位化 v' := v'/|v'|
def transform_direction(v: np.ndarray, R: np.ndarray, normalize: bool = True) -> np.ndarray:
    v = _ensure_3(v)
    R = _ensure_3x3(R)
    out = R @ v
    if normalize:
        n = np.linalg.norm(out)
        if n > 1e-10:
            out = out / n
    return out

# 多点变换: 对每点 p_i 做 p'_i = T*[p_i;1] 取前三维。即 P' = (T * P_h^T)^T 取前三列，P_h = [P, 1]
def transform_points(P: np.ndarray, T: np.ndarray) -> np.ndarray:
    P = np.asarray(P, dtype=np.float64)
    if P.ndim == 1:
        P = P.reshape(1, -1)
    T = _ensure_4x4(T)
    N = P.shape[0]
    ones = np.ones((N, 1))
    P_h = np.hstack([P[:, :3], ones])
    return (T @ P_h.T).T[:, :3]


# 位姿变换: origin' = T_B_A*origin；new_R = R_B_A @ R。公式: T_new = T_B_A * T_old（架 T_old = [R|origin; 0 0 0 1]）
def transform_pose(
    origin: np.ndarray, R: np.ndarray, T_B_A: np.ndarray
) -> Tuple[np.ndarray, np.ndarray]:
    origin = _ensure_3(origin)
    R = _ensure_3x3(R)
    T_B_A = _ensure_4x4(T_B_A)
    new_origin = transform_point(origin, T_B_A)
    R_B_A = T_B_A[:3, :3]
    new_R = R_B_A @ R
    return new_origin, new_R

# 完整架变换: T_new = T_B_A @ T_old（4×4 矩阵乘），同一刚体变换作用到架的原点与姿态
def transform_frame(T_old: np.ndarray, T_B_A: np.ndarray) -> np.ndarray:
    T_old = _ensure_4x4(T_old)
    T_B_A = _ensure_4x4(T_B_A)
    return T_B_A @ T_old

def transform_directions(V: np.ndarray, R: np.ndarray, normalize: bool = True) -> np.ndarray:
    V = np.asarray(V, dtype=np.float64)
    R = _ensure_3x3(R)
    if V.ndim == 1:
        V = V.reshape(1, -1)
    return (R @ V.T).T if not normalize else np.array([transform_direction(V[i], R, True) for i in range(V.shape[0])])

# 直线（一点+方向）变换: point' = T*point；direction' = R*direction 并单位化（方向仅旋转，不平移）
def transform_line(
    point: np.ndarray, direction: np.ndarray, T: np.ndarray
) -> Tuple[np.ndarray, np.ndarray]:
    point = _ensure_3(point)
    direction = _ensure_3(direction)
    T = _ensure_4x4(T)
    point_out = transform_point(point, T)
    direction_out = transform_direction(direction, T[:3, :3], True)
    return point_out, direction_out
